File: mbw/src/main/java/com/mycelium/wallet/activity/rmc/RMCAddressFragment.java
package com.mycelium.wallet.activity.rmc;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import com.google.common.base.Preconditions;
import com.mycelium.wallet.MbwManager;
import com.mycelium.wallet.R;
import com.mycelium.wallet.activity.rmc.adapter.AddressWidgetAdapter;
import com.mycelium.wallet.activity.view.ViewPagerIndicator;
import com.mycelium.wallet.event.AccountChanged;
import com.mycelium.wallet.event.BalanceChanged;
import com.mycelium.wallet.event.ReceivingAddressChanged;
import com.squareup.otto.Bus;
import com.squareup.otto.Subscribe;

import java.util.Calendar;
import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;


public class RMCAddressFragment extends Fragment {

    public static final String RMC_ACTIVE_PUSH_NOTIFICATION = "rmc_active_push_notification";
    private View _root;

    @BindView(R.id.active_in_day_progress)
    protected ProgressBar activeProgressBar;

    @BindView(R.id.active_in_day)
    protected TextView activeInDay;


    @BindView(R.id.view_pager)
    protected ViewPager viewPager;

    @BindView(R.id.title)
    protected TextView titleView;

    @BindView(R.id.pager_indicator)
    protected ViewPagerIndicator indicator;

    private MbwManager _mbwManager;
    private SharedPreferences sharedPreferences;
    private AddressWidgetAdapter adapter;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        _mbwManager = MbwManager.getInstance(getActivity());
        sharedPreferences = getActivity().getSharedPreferences("rmc_notification", Context.MODE_PRIVATE);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        _root = Preconditions.checkNotNull(inflater.inflate(R.layout.rmc_address_view, container, false));
        ButterKnife.bind(this, _root);
        return _root;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        adapter = new AddressWidgetAdapter(getActivity(), _mbwManager);
        viewPager.setAdapter(adapter);
        viewPager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                super.onPageSelected(position);
                titleView.setText(adapter.getPageTitle(position));
            }
        });
        indicator.setupWithViewPager(viewPager);
        viewPager.setCurrentItem(1);
        viewPager.postDelayed(new Runnable() {
            @Override
            public void run() {
                viewPager.setCurrentItem(0, true);
            }
        }, 3000);

        updateUi();
    }

    @OnClick(R.id.visit_rmc_one)
    void rmcOneClick() {
        startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://rmc.one")));
    }

    @Override
    public void onResume() {
        getEventBus().register(this);
        updateUi();
        super.onResume();
    }

    @Override
    public void onPause() {
        getEventBus().unregister(this);
        super.onPause();
    }

    private Bus getEventBus() {
        return _mbwManager.getEventBus();
    }

    private void activeBtnProgress() {
        Calendar calendarStart = Keys.getActiveStartDay();
        Calendar calendarEnd = Keys.getActiveEndDay();
        int progress = (int) TimeUnit.MILLISECONDS.toDays(Calendar.getInstance().getTimeInMillis() - calendarStart.getTimeInMillis());
        int total = (int) TimeUnit.MILLISECONDS.toDays(calendarEnd.getTimeInMillis() - calendarStart.getTimeInMillis());
        activeProgressBar.setProgress(progress);
        activeProgressBar.setMax(total);
        activeInDay.setText(getString(R.string.rmc_active_in_159_days, total - progress));
    }

    @OnClick(R.id.rmc_active_set_reminder)
    void setReminderClick() {
        final View view = LayoutInflater.from(getActivity()).inflate(R.layout.dialog_rmc_reminder, null, false);
        ((Switch) view.findViewById(R.id.add_push_notification)).setChecked(sharedPreferences.getBoolean(RMC_ACTIVE_PUSH_NOTIFICATION, false));
        new AlertDialog.Builder(getActivity())
                .setView(view)
                .setPositiveButton(R.string.save, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        if (((Switch) view.findViewById(R.id.add_to_calendar)).isChecked()) {
                            addEventToCalendar();
                        }
                        sharedPreferences.edit().putBoolean(RMC_ACTIVE_PUSH_NOTIFICATION
                                , ((Switch) view.findViewById(R.id.add_push_notification)).isChecked())
                                .apply();
                    }
                }).setNegativeButton(R.string.cancel, null)
                .create()
                .show();
    }

    private void addEventToCalendar() {

        Intent intent = new Intent(Intent.ACTION_EDIT);
        intent.setType("vnd.android.cursor.item/event");
        Calendar start = Keys.getActiveEndDay();
        long dtstart = start.getTimeInMillis();
        intent.putExtra("beginTime", dtstart);
        intent.putExtra("allDay", true);
        intent.putExtra("title", getString(R.string.rmc_activate));
        intent.putExtra("description", getString(R.string.rmc_activate_rmc));
        try {
            startActivity(intent);
        } catch (Exception ignore) {
            Toast.makeText(getActivity(), R.string.error_start_google_calendar, Toast.LENGTH_LONG).show();
        }
    }


    private void updateUi() {
        activeBtnProgress();
        adapter.notifyDataSetChanged();
    }

    @Subscribe
    public void receivingAddressChanged(ReceivingAddressChanged event) {
        updateUi();
    }

    @Subscribe
    public void accountChanged(AccountChanged event) {
        updateUi();
    }

    @Subscribe
    public void balanceChanged(BalanceChanged event) {
        updateUi();
    }
}

File: mbw/src/main/java/com/mycelium/wallet/activity/rmc/RMCAddressFragment.kt
package com.mycelium.wallet.activity.rmc

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Switch
import androidx.fragment.app.Fragment
import androidx.viewpager.widget.ViewPager.SimpleOnPageChangeListener
import com.mycelium.wallet.MbwManager
import com.mycelium.wallet.R
import com.mycelium.wallet.activity.modern.Toaster
import com.mycelium.wallet.activity.rmc.adapter.AddressWidgetAdapter
import com.mycelium.wallet.databinding.RmcAddressViewBinding
import com.mycelium.wallet.event.AccountChanged
import com.mycelium.wallet.event.BalanceChanged
import com.mycelium.wallet.event.ReceivingAddressChanged
import com.mycelium.wallet.external.partner.openLink
import com.squareup.otto.Bus
import com.squareup.otto.Subscribe
import java.util.Calendar
import java.util.concurrent.TimeUnit

class RMCAddressFragment : Fragment() {

    private var _mbwManager: MbwManager? = null
    private var sharedPreferences: SharedPreferences? = null
    private var adapter: AddressWidgetAdapter? = null
    private var binding: RmcAddressViewBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _mbwManager = MbwManager.getInstance(activity)
        sharedPreferences =
            activity!!.getSharedPreferences("rmc_notification", Context.MODE_PRIVATE)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = RmcAddressViewBinding.inflate(inflater).apply {
        binding = this
    }.root

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter = AddressWidgetAdapter(requireActivity(), _mbwManager!!)
        binding?.viewPager?.adapter = adapter
        binding?.viewPager!!.addOnPageChangeListener(object : SimpleOnPageChangeListener() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                binding?.title?.text = adapter!!.getPageTitle(position)
            }
        })
        binding?.pagerIndicator?.setupWithViewPager(binding?.viewPager!!)
        binding?.viewPager?.currentItem = 1
        binding?.viewPager?.postDelayed({ binding?.viewPager?.setCurrentItem(0, true) }, 3000)

        binding?.visitRmcOne?.setOnClickListener {
            openLink("https://rmc.one")
        }
        binding?.rmcActiveSetReminder?.setOnClickListener { setReminderClick() }
        updateUi()
    }


    override fun onResume() {
        eventBus.register(this)
        updateUi()
        super.onResume()
    }

    override fun onPause() {
        eventBus.unregister(this)
        super.onPause()
    }

    override fun onDestroyView() {
        binding = null
        super.onDestroyView()
    }

    private val eventBus: Bus
        get() = MbwManager.getEventBus()

    private fun activeBtnProgress() {
        val calendarStart = Keys.getActiveStartDay()
        val calendarEnd = Keys.getActiveEndDay()
        val progress =
            TimeUnit.MILLISECONDS.toDays(Calendar.getInstance().timeInMillis - calendarStart.timeInMillis)
                .toInt()
        val total =
            TimeUnit.MILLISECONDS.toDays(calendarEnd.timeInMillis - calendarStart.timeInMillis)
                .toInt()
        binding?.activeInDayProgress?.progress = progress
        binding?.activeInDayProgress?.max = total
        binding?.activeInDay?.text = getString(R.string.rmc_active_in_159_days, total - progress)
    }

    fun setReminderClick() {
        val view = LayoutInflater.from(activity).inflate(R.layout.dialog_rmc_reminder, null, false)
        (view.findViewById<View>(R.id.add_push_notification) as Switch).isChecked =
            sharedPreferences!!.getBoolean(
                RMC_ACTIVE_PUSH_NOTIFICATION,
                false
            )
        AlertDialog.Builder(activity)
            .setView(view)
            .setPositiveButton(
                R.string.save
            ) { dialogInterface, i ->
                if ((view.findViewById<View>(R.id.add_to_calendar) as Switch).isChecked) {
                    addEventToCalendar()
                }
                sharedPreferences!!.edit().putBoolean(
                    RMC_ACTIVE_PUSH_NOTIFICATION,
                    (view.findViewById<View>(R.id.add_push_notification) as Switch).isChecked
                )
                    .apply()
            }.setNegativeButton(R.string.cancel, null)
            .create()
            .show()
    }

    private fun addEventToCalendar() {
        val intent = Intent(Intent.ACTION_EDIT)
        intent.setType("vnd.android.cursor.item/event")
        val start = Keys.getActiveEndDay()
        val dtstart = start.timeInMillis
        intent.putExtra("beginTime", dtstart)
        intent.putExtra("allDay", true)
        intent.putExtra("title", getString(R.string.rmc_activate))
        intent.putExtra("description", getString(R.string.rmc_activate_rmc))
        try {
            startActivity(intent)
        } catch (ignore: Exception) {
            Toaster(this).toast(R.string.error_start_google_calendar, false)
        }
    }


    private fun updateUi() {
        activeBtnProgress()
        adapter!!.notifyDataSetChanged()
    }

    @Subscribe
    fun receivingAddressChanged(event: ReceivingAddressChanged?) {
        updateUi()
    }

    @Subscribe
    fun accountChanged(event: AccountChanged?) {
        updateUi()
    }

    @Subscribe
    fun balanceChanged(event: BalanceChanged?) {
        updateUi()
    }

    companion object {
        const val RMC_ACTIVE_PUSH_NOTIFICATION: String = "rmc_active_push_notification"
    }
}
